Codechange: Sprite sorting optimization

jmakovicka opened the pull request:
Another take on the sprite sorter optimization.

On top of the reverted 25ab9c1997f770f4a8a66bb3ad4b82ba87e3a977 ,
this patch adds the psdMax variable tracking the start of the
still intact pre-sorted portion of the parent sprite vector,
where the early bailout logic still works.

frosch123 committed:
Codechange: Sprite sorting optimization

Another take on the sprite sorter optimization.

On top of the reverted 25ab9c1997f770f4a8a66bb3ad4b82ba87e3a977 ,
this patch adds the psdMax variable tracking the start of the
still intact pre-sorted portion of the parent sprite vector,
where the early bailout logic still works.

jmakovicka commented:
This version of the 4k unzoom fix should avoid the sprite flickering caused by wrong sort order with the previous attempt, while maintaining the performance improvement.

michicc commented on src/viewport.cpp:
```diff
@@ -1404,11 +1405,23 @@ static bool ViewportSortParentSpritesChecker()
+	auto psdMax = psd;
+
+	/* pre-sort by xmin in ascending order */
+	QSortT(psdv->data(), psdv->size(), CompareParentSprites);
```
We are trying to move to something closer to modern C++, thus something with `std::sort` would be preferable.

PeterN commented on src/viewport.cpp:
```diff
@@ -1445,9 +1454,15 @@ static void ViewportSortParentSprites(ParentSpriteToSortVector *psdv)
+						break;
+					} else {
+						continue;
+					}
```
Can be simplified to:
`if (psd2 > psdMax) break;`
`continue;`

PeterN commented on src/viewport.cpp:
```diff
@@ -1458,6 +1473,8 @@ static void ViewportSortParentSprites(ParentSpriteToSortVector *psdv)
 			}
 			*psd = temp;
+			if (psd2 > psdMax)
+				psdMax = psd2;
```
OpenTTD-style is single-line or use brackets. This is fine here:
`if (psd2 > psdMax) psdMax = psd2;`

PeterN commented on src/viewport_sprite_sorter_sse4.cpp:
```diff
@@ -64,8 +75,14 @@ void ViewportSortParentSpritesSSE41(ParentSpriteToSortVector *psdv)
+					break;
+				} else {
+					continue;
+				}
```
Same as in viewport.cpp

PeterN commented on src/viewport_sprite_sorter_sse4.cpp:
```diff
@@ -89,6 +106,8 @@ void ViewportSortParentSpritesSSE41(ParentSpriteToSortVector *psdv)
 			}
 			*psd = temp;
+			if (psd2 > psdMax)
+				psdMax = psd2;
```
Same

PeterN commented:
This is still broken with regards to sorting, e.g. FIRS industry set.

jmakovicka committed:
sprite sorting regression test

print an exclamation mark if the output differs. input is presorted
for the original version too, as the exact result depends on the input.

jmakovicka commented:
This is weird. This time, I used https://github.com/jmakovicka/OpenTTD/commit/c96cc5a973c1f01638e1b94c9fedb20e38947b6f to test that the two versions give the same result, at least if the original is suipplied with the same pre-sorted input, as the exact sort result is input dependent.

PeterN commented:
The pre-sorting in your regression test messes up the sorting for the current sorter as well.

michicc commented:
Without having actually checked anything: Should this maybe use a stable sort (i.e. `std::stable_sort`)?
